File: server-py/app/registry.py
"""On-chain PairRegistry (X Layer 196) via web3.py: idempotent registration of
verified pairs plus the public read shape. Ports the Node registry.ts,
including the evidence-hash contract and the syncing lock with timeout."""
import asyncio
import json
import os
import time
from pathlib import Path

from web3 import Web3

CHAIN = 196
ARTIFACT = json.loads((Path(__file__).resolve().parents[2] / "contracts" / "artifacts" / "PairRegistry.json").read_text())
ABI = ARTIFACT["abi"]


def _rpc() -> str:
    return os.environ.get("XLAYER_RPC", "https://xlayerrpc.okx.com")


def _contract() -> str:
    return os.environ.get("REGISTRY_CONTRACT", "")


def _key() -> str:
    return os.environ.get("REGISTRY_OWNER_KEY", "")


def registry_configured() -> bool:
    return bool(Web3.is_address(_contract())) and bool(_key().startswith("0x") and len(_key()) == 66)


def evidence_json(r: dict) -> str:
    payload = {
        "block": r.get("block"),
        "chainId": str(r.get("chainId") or CHAIN),
        "checkedAt": r.get("checkedAt"),
        "liquidityUsd": r.get("liquidityUsd"),
        "meme": r["token"].lower(),
        "pool": r["pool"].lower(),
        "stock": r["stock"].lower(),
        "stockSide": r["stockSide"].lower(),
        "ticker": r.get("ticker"),
    }
    return json.dumps(payload, separators=(",", ":"), sort_keys=True, ensure_ascii=False)


def evidence_hash(r: dict) -> str:
    return Web3.keccak(text=evidence_json(r)).hex()


_w3: Web3 | None = None
_cache: tuple[float, list] | None = None


def w3() -> Web3:
    global _w3
    if _w3 is None:
        _w3 = Web3(Web3.HTTPProvider(_rpc(), request_kwargs={"timeout": 30}))
        if _w3.eth.chain_id != CHAIN:
            raise RuntimeError(f"wrong chain {_w3.eth.chain_id}, expect {CHAIN}")
    return _w3


async def onchain_pairs() -> list:
    global _cache
    contract = _contract()
    if not Web3.is_address(contract):
        return []
    if _cache and time.time() - _cache[0] < 60:
        return _cache[1]

    def read():
        c = w3().eth.contract(address=Web3.to_checksum_address(contract), abi=ABI)
        return [
            {
                "meme": p[0].lower(), "stock": p[1].lower(), "ticker": p[2],
                "evidence": p[3].hex(), "registeredAt": p[4],
            }
            for p in c.functions.all().call()
        ]

    rows = await asyncio.wait_for(asyncio.to_thread(read), 30)
    _cache = (time.time(), rows)
    return rows


async def registry_info(relations: list) -> dict:
    configured = registry_configured()
    try:
        rows = await onchain_pairs() if configured else []
    except Exception:
        rows = []
    done = {(p["meme"], p["stock"]) for p in rows}
    verified = [r for r in relations if r.get("status") == "verified"]
    by_pair = {(r["token"].lower(), r["stock"].lower()): r for r in verified}
    contract = _contract()
    return {
        "contract": contract or None,
        "chainId": CHAIN,
        "configured": configured,
        "explorer": f"https://www.okx.com/explorer/xlayer/address/{contract}" if contract else None,
        "onchainCount": len(rows),
        "pendingLocal": sum(1 for r in verified if (r["token"].lower(), r["stock"].lower()) not in done),
        "localVerified": len(verified),
        "pairs": [
            {**p, "evidenceSource": evidence_json(by_pair[(p["meme"], p["stock"])]) if (p["meme"], p["stock"]) in by_pair else None}
            for p in rows
        ],
    }


_syncing = False
_sync_started = 0.0


async def sync_registry(relations: list) -> None:
    global _syncing, _sync_started
    if not registry_configured():
        return
    if _syncing and time.time() - _sync_started < 900:
        return
    _syncing = True
    _sync_started = time.time()
    try:
        onchain = await onchain_pairs()
        done = {(p["meme"], p["stock"]) for p in onchain}
        pending = [r for r in relations if r.get("status") == "verified" and (r["token"].lower(), r["stock"].lower()) not in done]
        if not pending:
            return
        key = _key()
        account = w3().eth.account.from_key(key)
        if w3().eth.get_balance(account.address) == 0:
            print("[registry] owner wallet empty, skip sync")
            return
        contract = w3().eth.contract(address=Web3.to_checksum_address(_contract()), abi=ABI)

        def register_all():
            for r in pending:
                try:
                    tx = contract.functions.register(
                        Web3.to_checksum_address(r["token"].lower()),
                        Web3.to_checksum_address(r["stock"].lower()),
                        r.get("ticker", ""),
                        bytes.fromhex(evidence_hash(r)[2:]),
                    ).build_transaction({
                        "from": account.address,
                        "nonce": w3().eth.get_transaction_count(account.address),
                        "chainId": CHAIN,
                        "gas": 200_000,
                    })
                    tx["gasPrice"] = w3().eth.gas_price
                    signed = account.sign_transaction(tx)
                    h = w3().eth.send_raw_transaction(signed.raw_transaction)
                    w3().eth.wait_for_transaction_receipt(h, timeout=120)
                    print(f"[registry] registered {r.get('ticker')} {r['token'][:10]} tx={h.hex()}")
                except Exception as e:
                    print(f"[registry] register failed {r.get('ticker')}: {e}")

        await asyncio.wait_for(asyncio.to_thread(register_all), 60 * len(pending) + 60)
        global _cache
        _cache = None
    finally:
        _syncing = False

File: src/lib/registry.ts
// On-chain PairRegistry (X Layer mainnet): writes verified meme×stock pairs,
// reads them back for the public API. The evidence JSON is hashed with
// keccak256 before upload so anyone can re-derive it from /api/registry.
import { ethers } from 'ethers';
import { readFileSync } from 'node:fs';
import type { XRelation } from './xlayer';

const artifact = JSON.parse(readFileSync(new URL('../../contracts/artifacts/PairRegistry.json', import.meta.url), 'utf8')) as { abi: any[]; bytecode: string };
const RPC = () => process.env.XLAYER_RPC ?? 'https://xlayerrpc.okx.com';
// Read env lazily: ESM imports run before server.ts calls loadEnvFile.
const CONTRACT = () => process.env.REGISTRY_CONTRACT ?? '';
const KEY = () => process.env.REGISTRY_OWNER_KEY ?? '';
const CHAIN = 196;

export function registryConfigured() {
  return /^0x[0-9a-fA-F]{40}$/.test(CONTRACT()) && /^0x[0-9a-fA-F]{64}$/.test(KEY());
}

// Keys are ASCII-ordered literals so JSON.stringify output is deterministic.
export function evidenceJson(r: XRelation) {
  return JSON.stringify({
    block: r.block,
    chainId: String(r.chainId ?? CHAIN),
    checkedAt: r.checkedAt,
    liquidityUsd: r.liquidityUsd ?? null,
    meme: r.token.toLowerCase(),
    pool: r.pool.toLowerCase(),
    stock: r.stock.toLowerCase(),
    stockSide: r.stockSide.toLowerCase(),
    ticker: r.ticker,
  });
}
export function evidenceHash(r: XRelation) {
  return ethers.keccak256(ethers.toUtf8Bytes(evidenceJson(r)));
}

let provider: ethers.JsonRpcProvider | null = null;
let cached: { at: number; rows: any[] } | null = null;

function readOnly() {
  provider ??= new ethers.JsonRpcProvider(RPC(), CHAIN, { staticNetwork: true });
  return new ethers.Contract(CONTRACT(), artifact.abi, provider);
}

export async function onchainPairs(): Promise<any[]> {
  if (!/^0x[0-9a-fA-F]{40}$/.test(CONTRACT())) return [];
  if (cached && Date.now() - cached.at < 60_000) return cached.rows;
  const raw = await readOnly().all();
  const rows = raw.map((p: any) => ({
    meme: p.meme.toLowerCase(), stock: p.stock.toLowerCase(), ticker: p.ticker,
    evidence: p.evidence, registeredAt: Number(p.registeredAt),
  }));
  cached = { at: Date.now(), rows };
  return rows;
}

export async function registryInfo(relations: XRelation[]) {
  const configured = registryConfigured();
  const rows = configured ? await onchainPairs().catch(() => []) : [];
  const known = new Set(relations.map(r => `${r.token.toLowerCase()}:${r.stock.toLowerCase()}`));
  return {
    contract: CONTRACT() || null, chainId: CHAIN, configured,
    explorer: CONTRACT() ? `https://www.okx.com/explorer/xlayer/address/${CONTRACT()}` : null,
    onchainCount: rows.length,
    pendingLocal: relations.filter(r => r.status === 'verified' && !rows.some(p => p.meme === r.token.toLowerCase() && p.stock === r.stock.toLowerCase())).length,
    localVerified: relations.filter(r => r.status === 'verified').length,
    pairs: rows.map(p => ({ ...p, evidenceSource: relations.find(r => r.token.toLowerCase() === p.meme && r.stock.toLowerCase() === p.stock && r.status === 'verified') ? evidenceJson(relations.find(r => r.token.toLowerCase() === p.meme && r.stock.toLowerCase() === p.stock)!) : null })),
    knownUntracked: rows.filter(p => !known.has(`${p.meme}:${p.stock}`)).length,
  };
}

// ethers waits have no built-in timeout; a hung RPC call would hold the
// syncing lock forever. Recover by allowing a fresh loop round after 15 min.
const withTimeout = <T>(p: Promise<T>, ms: number) => Promise.race([p, new Promise<T>((_, reject) => setTimeout(() => reject(new Error('registry rpc timeout')), ms))]);
let syncing = false, syncStartedAt = 0;
export async function syncRegistry(relations: XRelation[]) {
  if (!registryConfigured()) return;
  if (syncing && Date.now() - syncStartedAt < 900_000) return;
  syncing = true; syncStartedAt = Date.now();
  try {
    const onchain = await onchainPairs();
    const done = new Set(onchain.map(p => `${p.meme}:${p.stock}`));
    const pending = relations.filter(r => r.status === 'verified' && !done.has(`${r.token.toLowerCase()}:${r.stock.toLowerCase()}`));
    if (!pending.length) return;
    const wallet = new ethers.Wallet(KEY(), provider ??= new ethers.JsonRpcProvider(RPC(), CHAIN, { staticNetwork: true }));
    const balance = await wallet.provider!.getBalance(wallet.address);
    if (balance === 0n) { console.error('[registry] owner wallet empty, skip sync'); return; }
    const contract = new ethers.Contract(CONTRACT(), artifact.abi, wallet);
    for (const r of pending) {
      try {
        const tx = await withTimeout(contract.register(r.token.toLowerCase(), r.stock.toLowerCase(), r.ticker, evidenceHash(r)), 60_000);
        await withTimeout(tx.wait(), 120_000);
        console.log(`[registry] registered ${r.ticker} ${r.token.slice(0, 10)} tx=${tx.hash}`);
      } catch (e) {
        console.error(`[registry] register failed ${r.ticker}: ${e instanceof Error ? e.message : e}`);
      }
    }
    cached = null;
  } finally {
    syncing = false;
  }
}
